File: backend/src/api/callbacks.rs
use axum::{Json, extract::{State, Path}};
use serde_json::Value;
use std::sync::Arc;
use crate::{app::AppState, db::queries, error::{AppResult, AppError}, api::events::SseEvent, notifications, notifications::telegram::TelegramEvent};

pub async fn progress_callback(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let phase = body.get("phase").and_then(|v| v.as_str()).unwrap_or("download");
    let progress_pct = body.get("progress_pct").and_then(|v| v.as_i64()).unwrap_or(0) as i32;

    queries::update_job_progress(&state.db, &id, phase, progress_pct).await?;
    queries::update_job_phase(&state.db, &id, phase).await?;

    // Insert chunk info if present
    if let Some(chunk) = body.get("chunk") {
        if let (Some(filename), Some(discord_url)) = (
            chunk.get("filename").and_then(|v| v.as_str()),
            chunk.get("discord_url").and_then(|v| v.as_str()),
        ) {
            let chunk_idx = chunk.get("chunk_index").and_then(|v| v.as_i64()).unwrap_or(0);
            let duration = chunk.get("duration_seconds").and_then(|v| v.as_f64());
            queries::insert_hls_chunk(&state.db, &queries::NewHlsChunk {
                job_id: id.clone(),
                chunk_index: chunk_idx,
                filename: filename.to_string(),
                discord_url: Some(discord_url.to_string()),
                discord_message_id: chunk.get("discord_message_id").and_then(|v| v.as_str()).map(String::from),
                duration_seconds: duration,
                file_size_bytes: None,
            }).await?;
        }
    }

    // Log event
    queries::insert_job_event(
        &state.db, &id, Some(phase), "progress",
        &format!("Progress: {}%", progress_pct), Some(progress_pct as i64),
    ).await?;

    // Broadcast
    let _ = state.event_tx.send(SseEvent::JobProgress {
        job_id: id,
        phase: phase.to_string(),
        progress_pct,
    });

    Ok(Json(serde_json::json!({ "ok": true })))
}

pub async fn checkpoint_callback(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let checkpoint = body.get("checkpoint")
        .and_then(|v| v.as_str())
        .ok_or_else(|| AppError::BadRequest("Missing checkpoint field".into()))?;
    let artifact_id = body.get("artifact_id")
        .and_then(|v| v.as_str());
    let file_url = body.get("file_url")
        .and_then(|v| v.as_str());

    queries::update_job_checkpoint(&state.db, &id, checkpoint, artifact_id, file_url).await?;

    queries::insert_job_event(
        &state.db, &id, Some(checkpoint), "checkpoint",
        &format!("Checkpoint saved: {}", checkpoint), None,
    ).await?;

    let _ = state.event_tx.send(SseEvent::JobCheckpoint {
        job_id: id.clone(),
        checkpoint: checkpoint.to_string(),
    });

    // Telegram notification
    let title = queries::get_job(&state.db, &id).await
        .ok()
        .and_then(|j| j.title)
        .unwrap_or_default();
    notifications::send_notification(&state, TelegramEvent::CheckpointSaved(
        title, checkpoint.to_string(),
    ));

    Ok(Json(serde_json::json!({ "ok": true })))
}

pub async fn complete_callback(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let resolution = body.get("video_resolution")
        .and_then(|v| v.as_str())
        .unwrap_or("1080p");
    let duration = body.get("duration_seconds")
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0);

    // Get job before updating to capture gh_run_id
    let job = queries::get_job(&state.db, &id).await.ok();
    let gh_run_id = job.as_ref().and_then(|j| j.gh_run_id.clone());

    queries::update_job_completed(&state.db, &id, resolution, duration).await?;

    queries::insert_job_event(
        &state.db, &id, None, "status_change",
        &format!("Completed: {} resolution, {:.0}s duration", resolution, duration), None,
    ).await?;

    let _ = state.event_tx.send(SseEvent::JobCompleted {
        job_id: id.clone(),
    });

    // Telegram notification
    let title = job.as_ref().and_then(|j| j.title.clone()).unwrap_or_default();
    let details = format!("{} resolution, {:.0}s duration", resolution, duration);
    notifications::send_notification(&state, TelegramEvent::JobCompleted(title, details));

    // Delete GitHub Actions run if configured
    if let Some(run_id) = gh_run_id {
        let config = state.config.read().await;
        if let (Some(token), Some(repo)) = (&config.gh_token, &config.gh_repo) {
            let token = token.clone();
            let repo = repo.clone();
            drop(config);
            if let Some((owner, repo_name)) = repo.split_once('/') {
                let url = format!("https://api.github.com/repos/{}/{}/actions/runs/{}", owner, repo_name, run_id);
                let _ = state.http.delete(&url)
                    .bearer_auth(&token)
                    .header("Accept", "application/vnd.github+json")
                    .header("X-GitHub-Api-Version", "2022-11-28")
                    .send()
                    .await;
                queries::insert_job_event(
                    &state.db, &id, None, "cleanup",
                    &format!("GitHub Actions run {} deleted", run_id), None,
                ).await.ok();
            }
        }
    }

    Ok(Json(serde_json::json!({ "ok": true })))
}

pub async fn failed_callback(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let error_msg = body.get("error_message")
        .and_then(|v| v.as_str())
        .unwrap_or("Unknown error");

    queries::update_job_failed(&state.db, &id, error_msg).await?;

    queries::insert_job_event(
        &state.db, &id, None, "error",
        &format!("Failed: {}", error_msg), None,
    ).await?;

    let _ = state.event_tx.send(SseEvent::JobFailed {
        job_id: id.clone(),
        error: error_msg.to_string(),
    });

    // Telegram notification
    let job = queries::get_job(&state.db, &id).await.ok();
    let title = job.as_ref().and_then(|j| j.title.clone()).unwrap_or_default();
    let phase = job.as_ref().and_then(|j| j.current_phase.clone()).unwrap_or_default();
    notifications::send_notification(&state, TelegramEvent::JobFailed(
        title, phase, error_msg.to_string(),
    ));

    Ok(Json(serde_json::json!({ "ok": true })))
}

File: backend-bun/src/api/callbacks.ts
import type { Context } from "hono";
import type { AppBindings } from "../app";
import { badRequest } from "../error";
import * as queries from "../db/queries";
import { sendNotification } from "../notifications/telegram";

export async function progressCallback(c: Context<AppBindings>): Promise<Response> {
  const id = c.req.param("id") as string;
  const body = await c.req.json<Record<string, any>>();
  const phase: string = body.phase ?? "download";
  const progressPct: number = body.progress_pct ?? 0;

  queries.updateJobProgress(c.var.db, id, phase, progressPct);
  queries.updateJobPhase(c.var.db, id, phase);

  // Insert HLS chunk info if present
  if (body.chunk) {
    queries.insertHlsChunk(c.var.db, {
      jobId: id,
      chunkIndex: body.chunk.chunk_index ?? 0,
      filename: body.chunk.filename,
      discordUrl: body.chunk.discord_url ?? null,
      discordMessageId: body.chunk.discord_message_id ?? null,
      durationSeconds: body.chunk.duration_seconds ?? null,
      fileSizeBytes: null,
    });
  }

  // Log event
  queries.insertJobEvent(c.var.db, id, phase, "progress", `Progress: ${progressPct}%`, progressPct);

  // Broadcast
  c.var.eventBus.send({ type: "job_progress", data: { job_id: id, phase, progress_pct: progressPct } });

  return c.json({ ok: true });
}

export async function checkpointCallback(c: Context<AppBindings>): Promise<Response> {
  const id = c.req.param("id") as string;
  const body = await c.req.json<Record<string, any>>();
  if (!body.checkpoint) throw badRequest("Missing checkpoint field");

  const artifactId = body.artifact_id ?? null;
  const fileUrl = body.file_url ?? null;

  queries.updateJobCheckpoint(c.var.db, id, body.checkpoint, artifactId, fileUrl);

  // Log event
  queries.insertJobEvent(c.var.db, id, body.checkpoint, "checkpoint", `Checkpoint saved: ${body.checkpoint}`, null);

  // Broadcast
  c.var.eventBus.send({ type: "job_checkpoint", data: { job_id: id, checkpoint: body.checkpoint } });

  // Telegram notification
  const job = queries.getJob(c.var.db, id);
  const title = job?.title ?? "Unknown";
  sendNotification(c, { type: "CheckpointSaved", title, phase: body.checkpoint });

  return c.json({ ok: true });
}

export async function completeCallback(c: Context<AppBindings>): Promise<Response> {
  const id = c.req.param("id") as string;
  const body = await c.req.json<Record<string, any>>();
  const resolution = body.video_resolution || "1080p";
  const duration = body.duration_seconds ?? 0;

  console.log(`[callback] complete job=${id} resolution=${resolution} duration=${duration}`);

  try {
    // Get job before update to capture gh_run_id
    const job = queries.getJob(c.var.db, id);
    const ghRunId = job.ghRunId;

    queries.updateJobCompleted(c.var.db, id, resolution, duration);
    console.log(`[callback] job ${id} marked completed`);

    // Log event
    queries.insertJobEvent(c.var.db, id, null, "status_change", `Completed: ${resolution}, ${duration}s duration`, null);

    // Broadcast
    c.var.eventBus.send({ type: "job_completed", data: { job_id: id } });

    // Telegram notification
    const title = job?.title ?? "Unknown";
    sendNotification(c, { type: "JobCompleted", title, details: `${resolution}, ${duration}s duration` });

    // Clean up GHA run
    if (ghRunId) {
      const ghToken = queries.getSetting(c.var.db, "gh_token");
      const ghRepo = queries.getSetting(c.var.db, "gh_repo");
      if (ghToken && ghRepo) {
        const url = `https://api.github.com/repos/${ghRepo}/actions/runs/${ghRunId}`;
        fetch(url, {
          method: "DELETE",
          headers: {
            Authorization: `Bearer ${ghToken}`,
            Accept: "application/vnd.github+json",
          },
        }).catch((e) => console.error(`[callback] GHA cleanup failed job=${id}:`, e));
      }
    }
  } catch (e: any) {
    console.error(`[callback] complete failed job=${id}:`, e);
    return c.json({ error: e.message || "Internal error" }, 500);
  }

  return c.json({ ok: true });
}

export async function failedCallback(c: Context<AppBindings>): Promise<Response> {
  const id = c.req.param("id") as string;
  const body = await c.req.json<Record<string, any>>();
  const errorMsg = body.error_message || "Unknown error";

  queries.updateJobFailed(c.var.db, id, errorMsg);

  // Log event
  queries.insertJobEvent(c.var.db, id, null, "error", `Failed: ${errorMsg}`, null);

  // Broadcast
  c.var.eventBus.send({ type: "job_failed", data: { job_id: id, error: errorMsg } });

  // Telegram notification
  const job = queries.getJob(c.var.db, id);
  const title = job?.title ?? "Unknown";
  const phase = job?.currentPhase ?? "unknown";
  sendNotification(c, { type: "JobFailed", title, phase, error: errorMsg });

  return c.json({ ok: true });
}
